ramanujan/matrix.py

```diff
@@ -48,21 +48,42 @@ def walk(self, trajectory, iterations, start):
         where `M=self`, `(t_0, ..., t_k)=trajectory`, `n=iterations` and `(s_0, ..., s_k)=start`.
 
         This is a generalization of the basic (and most common) case $\prod_{i=0}^{n-1}M(s+i)$,
-        where `M=self`, `n=iterations` and `s=start`.
+        where M=self, n=iterations and s=start.
 
         Args:
             trajectory: the trajectory of a single step in the walk, as defined above.
-            iterations: the amount of multiplications to perform
+            iterations: The amount of multiplications to perform or a list of integers representing indexes along the multipication proccess to be returned. 
+                        If it's a list, the iterations will be sorted in ascending order to improve efficiency. 
             start: the starting point of the matrix multiplication
         Returns:
-            the walk multiplication as defined above.
+            Steps from the walk multiplication as defined above.
         """
+
         position = start
-        retval = Matrix.eye(2)
-        for _ in range(iterations):
-            retval *= self.subs(position)
+        value_matrix = Matrix.eye(2)
+
+        retval = [value_matrix]
```

Comment:
This (first updating and then updating) is confusing, why not just append after the update?